src/EditorFeatures/CSharpTest/CodeActions/SyncNamespace/SyncNamespaceTests_MoveFile.cs

```diff
@@ -0,0 +1,310 @@
+        <Document Folders=""{documentPath.folder}"" FilePath=""{documentPath.filePath}""> 
+namespace [||]{declaredNamespace}
+{{    
+    class Class1
+    {{
+    }}
+}}  
+        </Document>
+    </Project>
+</Workspace>";
+            await TestMoveFileToMatchNamespace(code, expectedFolders);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CodeActionsSyncNamespace)]
+        public async Task MoveFile_SingleAction1()
+        {
+            // current path is <root>\
+            // expected new path is <root>\B\C\
+
+            var defaultNamespace = "A";
+            var declaredNamespace = "A.B.C";
+
+            var expectedFolders = new List<string[]>();
+            expectedFolders.Add(new[] { "B", "C" });
+
+            var documentPath = CreateDocumentFilePath(Array.Empty<string>());
+            var code =
+$@"
+<Workspace>
+    <Project Language=""C#"" AssemblyName=""Assembly1"" FilePath=""{ProjectFilePath}"" DefaultNamespace=""{defaultNamespace}"" CommonReferences=""true"">
+        <Document Folders=""{documentPath.folder}"" FilePath=""{documentPath.filePath}""> 
+namespace [||]{declaredNamespace}
+{{    
+    class Class1
+    {{
+    }}
+}}  
+        </Document>
+    </Project>
+</Workspace>";
+            await TestMoveFileToMatchNamespace(code, expectedFolders);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CodeActionsSyncNamespace)]
+        public async Task MoveFile_SingleAction2()
+        {
+            // current path is <root>\
+            // expected new path is <root>\B\C\D\E\
+
+            var defaultNamespace = "A";
+            var declaredNamespace = "A.B.C.D.E";
+
+            var expectedFolders = new List<string[]>();
+            expectedFolders.Add(new[] { "B", "C", "D", "E" });
+
+            var documentPath1 = CreateDocumentFilePath(Array.Empty<string>(), "File1.cs");
+            var documentPath2 = CreateDocumentFilePath(new[] { "B", "C" }, "File2.cs");   // file2 is in <root>\B\C\
+            var code =
+$@"
+<Workspace>
+    <Project Language=""C#"" AssemblyName=""Assembly1"" FilePath=""{ProjectFilePath}"" DefaultNamespace=""{defaultNamespace}"" CommonReferences=""true"">
+        <Document Folders=""{documentPath1.folder}"" FilePath=""{documentPath1.filePath}""> 
+namespace [||]{declaredNamespace}
+{{    
+    class Class1
+    {{
+    }}
+}}  
+        </Document>        
+        <Document Folders=""{documentPath2.folder}"" FilePath=""{documentPath2.filePath}""> 
+namespace Foo
+{{    
+    class Class2
+    {{
+    }}
+}}  
+        </Document>  
+    </Project>
+</Workspace>";
+            await TestMoveFileToMatchNamespace(code, expectedFolders);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CodeActionsSyncNamespace)]
+        public async Task MoveFile_MoveToRoot()
+        {
+            // current path is <root>\A\B\C\
+            // expected new path is <root>
+
+            var defaultNamespace = "";
```

Comment:
so, this make sure that this refactoring can't never put file outside of repo root or something? also, handle access deny thing properly? (I mean any io exception). by handle properly, I mean not crash.

Reply:
Yes, any file movement must be under project root, and the refactoring won't trigger on file outside of project root.

In terms of exception, I'm mostly relying on VS workspace to handle it. Is that sufficient?